File: iosApp/iosApp/ui/game/GameScreen.swift
//
//  GameScreen.swift
//  iosApp
//
//  Created by Fakhri Rasyid Saputro on 23/01/25.
//  Copyright © 2025 orgName. All rights reserved.
//

import SwiftUI
import Combine
import shared

struct GameScreen: View {
    @EnvironmentObject var router: Router
    
    @StateObject var viewModel: GameViewModel
    @State private var lastDragValue: CGFloat = 0
    
    @State private var showPauseDialog: Bool = false
    
    init(playerName: String, gameDifficulty: String) {
        _viewModel = StateObject(wrappedValue: GameViewModel(playerName: playerName, gameDifficulty: gameDifficulty))
    }

    var body: some View {
        GeometryReader { geometry in
            ZStack {
                // Scrolling Background                
                ScrollingStarsBackground(
                    isPaused: ($viewModel.isPaused.wrappedValue || $viewModel.isButtonPaused.wrappedValue || $viewModel.isGameEnded.wrappedValue),
                    scrollSpeed: 1
                )
            
                // Player
                Image(resource: \.asset_player_point)
                    .resizable()
                    .frame(width: viewModel.playerWidth, height: viewModel.playerHeight)
                    .position(
                        x: viewModel.playerPosition + (viewModel.playerWidth / 2),
                        y: geometry.size.height - viewModel.paddingToDrawPlayerX
                    )
                    .shadow(
                        color: Color(SharedRes.colors().colorOnPrimary.getUIColor()),
                        radius: 10,
                        x: 0,
                        y: 0
                    )

                // Bullets
                ForEach(Array(viewModel.bullets.enumerated()), id: \.offset) { index, bullet in
                    Image(resource: \.asset_player_bullet_point)
                        .resizable()
                        .frame(width: viewModel.bulletWidth, height: viewModel.bulletHeight)
                        .position(bullet)
                }


                // Enemies
                ForEach(viewModel.enemies, id: \.self) { enemy in
                    Image(resource: enemy.isRed ? \.asset_enemy_red_one_point : \.asset_enemy_green_one_point)
                        .resizable()
                        .frame(width: viewModel.enemyWidth, height: viewModel.enemyHeight)
                        .position(x: enemy.position.x + (viewModel.enemyWidth / 2), y: enemy.position.y)
                        .shadow(
                            color: enemy.isRed ? Color(SharedRes.colors().primaryRed.getUIColor()) : Color(SharedRes.colors().primaryGreen.getUIColor()),
                            radius: 10,
                            x: 0,
                            y: 0
                        )
                }

                // Enemy Bullets
                ForEach(Array(viewModel.enemyBullets.enumerated()), id: \.offset) { index, bullet in
                    Image(resource: (viewModel.gameDifficulty == GameDifficulty.EASY.name) ? \.asset_enemy_bullet_point_green : \.asset_enemy_bullet_point_red)
                        .resizable()
                        .frame(width: viewModel.bulletWidth, height: viewModel.bulletHeight)
                        .position(bullet)
                }


                // Score and Lives
                VStack {
                    HStack {
                        Text(
                            String(
                                format: SharedRes.strings().game_score.localized(),
                                viewModel.score
                            )
                        ).font(Font(resource: \.pressstart2p_regular, withSize: 14.0))
                            .foregroundStyle(.white)
                        
                        Spacer()
                        
                        Button(action: {
                            viewModel.onButtonPause()
                            showPauseDialog = true
                        }) {
                            Image(resource: \.icon_pause)
                                .resizable()
                                .frame(width: 32, height: 32)
                        }
                        
                        Spacer()
                        
                        Text(
                            String(
                                format: SharedRes.strings().game_lives.localized(),
                                viewModel.lives
                            )
                        ).font(Font(resource: \.pressstart2p_regular, withSize: 14.0))
                            .foregroundStyle(.white)
                    }
                    .padding()
                    .padding(.top, 42)
                    Spacer()
                }
                
                if showPauseDialog {
                    GamePauseDialog(
                        isPresented: $showPauseDialog,
                        isDisabled: true,
                        onResumeSelected: { viewModel.onButtonResume() },
                        onEndGameSelected: { viewModel.stopGame() },
                        onLeaveGameSelected: { self.router.navigateBack() }
                    )
                }
                
                if viewModel.isGameEnded {
                    GameEndedDialog(
                        isPresented: $viewModel.isGameEnded,
                        isDisabled: true,
                        playerResult: viewModel.playerResult,
                        onLeaderboardsSelected: { self.router.navigate(to: .LEADERBOARDS) },
                        onLeaveGameSelected: { self.router.navigateBack() }
                    )
                }
            }
            .onAppear {
                viewModel.screenWidth = geometry.size.width
                viewModel.screenHeight = geometry.size.height
                viewModel.startGameLoops()
            }
            .gesture(
                DragGesture()
                    .onChanged { value in
                        let delta = (value.translation.width - lastDragValue)
                        viewModel.movePlayer(delta: delta)
                        lastDragValue = value.translation.width
                    }
                    .onEnded { _ in
                        lastDragValue = 0
                    }
            )
            .onReceive(NotificationCenter.default.publisher(for: UIApplication.willEnterForegroundNotification)) { _ in
                viewModel.onResume()
            }
            .onReceive(NotificationCenter.default.publisher(for: UIApplication.willResignActiveNotification)) { _ in
                viewModel.onPause()
            }
        }
    }
}

File: androidApp/src/main/java/com/fakhrirasyids/spaceattack/android/ui/screens/game/GameScreen.kt
package com.fakhrirasyids.spaceattack.android.ui.screens.game

import androidx.compose.foundation.Image
import androidx.compose.foundation.gestures.detectHorizontalDragGestures
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.Lifecycle
import com.fakhrirasyids.spaceattack.SharedRes
import com.fakhrirasyids.spaceattack.android.ui.screens.game.widgets.GameEndedDialog
import com.fakhrirasyids.spaceattack.android.ui.screens.game.widgets.GamePauseDialog
import com.fakhrirasyids.spaceattack.android.ui.widgets.ScrollingStarsBackground
import com.fakhrirasyids.spaceattack.android.utils.enums.GameDifficulty
import com.fakhrirasyids.spaceattack.android.utils.helpers.ComposableLifecycle
import com.fakhrirasyids.spaceattack.android.utils.helpers.DensityConverter.asDp
import com.fakhrirasyids.spaceattack.android.utils.helpers.DensityConverter.asPx
import com.fakhrirasyids.spaceattack.android.utils.helpers.advanceShadow
import dev.icerock.moko.resources.compose.fontFamilyResource
import dev.icerock.moko.resources.compose.painterResource
import dev.icerock.moko.resources.compose.stringResource

@Composable
fun GameScreen(
    modifier: Modifier = Modifier,
    gameViewModel: GameViewModel,
    navigateBack: () -> Unit
) {
    val context = LocalContext.current

    var showPauseDialog by remember { mutableStateOf(false) }

    val enemyType = gameViewModel.difficulty
    val paddingToDrawPlayerX = 200

    val playerWidthDp = gameViewModel.playerWidthDp.asDp
    val playerHeightDp = gameViewModel.playerHeightDp.asDp

    val enemyWidthDp = gameViewModel.enemyWidthDp.asDp
    val enemyHeightDp = gameViewModel.enemyHeightDp.asDp

    val bulletSizeDp = gameViewModel.bulletSizeDp.asDp

    val playerWidthPx = gameViewModel.playerWidthDp.asPx

    val isPaused by gameViewModel.isPaused.collectAsState()
    val isButtonPaused by gameViewModel.isButtonPaused.collectAsState()
    val isGameEnded by gameViewModel.isGameEnded.collectAsState()

    val playerResult by gameViewModel.playerResult.collectAsState()

    val playerX by gameViewModel.playerPosition.collectAsState()
    val bullets by gameViewModel.bullets.collectAsState()
    val enemies by gameViewModel.enemies.collectAsState()
    val enemyBullets by gameViewModel.enemyBullets.collectAsState()
    val score by gameViewModel.score.collectAsState()
    val lives by gameViewModel.lives.collectAsState()

    val screenWidth by gameViewModel.screenWidth.collectAsState()
    val screenHeight by gameViewModel.screenHeight.collectAsState()

    ComposableLifecycle { _, event ->
        when (event) {
            Lifecycle.Event.ON_PAUSE -> gameViewModel.onPause()
            Lifecycle.Event.ON_RESUME -> gameViewModel.onResume()
            else -> {}
        }
    }

    LaunchedEffect(screenWidth, screenHeight) {
        if (screenWidth > 0 && screenHeight > 0) {
            gameViewModel.startGameLoops()
        }
    }

    Box(
        modifier = modifier
            .fillMaxSize()
            .onGloballyPositioned { layoutCoordinates ->
                gameViewModel.screenWidth.value = layoutCoordinates.size.width
                gameViewModel.screenHeight.value = layoutCoordinates.size.height
            }
            .pointerInput(Unit) {
                detectHorizontalDragGestures { _, dragAmount ->
                    gameViewModel.movePlayer(dragAmount.toInt(), screenWidth, playerWidthPx)
                }
            }
    ) {
        ScrollingStarsBackground(
            isPaused = isPaused || isButtonPaused || isGameEnded,
            modifier = Modifier.fillMaxSize()
        )

        // Player
        Image(
            contentScale = ContentScale.FillBounds,
            modifier = Modifier
                .offset {
                    IntOffset(
                        playerX,
                        screenHeight - paddingToDrawPlayerX
                    )
                }
                .size(width = playerWidthDp, height = playerHeightDp)
                .advanceShadow(
                    color = Color(SharedRes.colors.colorOnPrimary.getColor(context))
                ),
            painter = painterResource(imageResource = SharedRes.images.asset_player_point),
            contentDescription = null
        )

        // Player Bullets
        bullets.forEach { bullet ->
            Image(
                modifier = Modifier
                    .offset { IntOffset(bullet.x.toInt(), bullet.y.toInt()) }
                    .size(bulletSizeDp),
                painter = painterResource(
                    imageResource = SharedRes.images.asset_player_bullet_point
                ),
                contentDescription = null
            )
        }

        // Enemies
        enemies.forEach { enemy ->
            Image(
                modifier = Modifier
                    .offset { IntOffset(enemy.position.x.toInt(), enemy.position.y.toInt()) }
                    .size(width = enemyWidthDp, height = enemyHeightDp)
                    .advanceShadow(
                        color = if (enemy.isRed) Color(
                            SharedRes.colors.primaryRed.getColor(
                                context
                            )
                        ) else Color(SharedRes.colors.primaryGreen.getColor(context)),
                    ),
                painter = painterResource(
                    imageResource = if (enemy.isRed) SharedRes.images.asset_enemy_red_one_point else SharedRes.images.asset_enemy_green_one_point
                ),
                contentDescription = null
            )
        }

        // Enemy Bullets
        enemyBullets.forEach { bullet ->
            Image(
                modifier = Modifier
                    .offset { IntOffset(bullet.x.toInt(), bullet.y.toInt()) }
                    .size(bulletSizeDp),
                painter = painterResource(
                    imageResource = if (enemyType == GameDifficulty.Hard.name) SharedRes.images.asset_enemy_bullet_point_red else SharedRes.images.asset_enemy_bullet_point_green
                ),
                contentDescription = null
            )
        }

        // Score and Lives
        Row(modifier = Modifier.padding(16.dp)) {
            Text(
                text = stringResource(SharedRes.strings.game_score, score),
                fontSize = 12.sp,
                fontFamily = fontFamilyResource(SharedRes.fonts.pressstart2p_regular),
                color = Color.White
            )

            IconButton(
                onClick = {
                    gameViewModel.setButtonPause()
                    showPauseDialog = true
                },
                modifier = Modifier
                    .weight(1f)
            ) {
                Icon(
                    painter = painterResource(SharedRes.images.icon_pause),
                    contentDescription = stringResource(SharedRes.strings.app_name),
                    tint = Color.White
                )
            }

            Text(
                text = stringResource(SharedRes.strings.game_lives, lives),
                fontSize = 12.sp,
                fontFamily = fontFamilyResource(SharedRes.fonts.pressstart2p_regular),
                color = Color.White
            )
        }

        if (showPauseDialog) {
            GamePauseDialog(
                onDismiss = { showPauseDialog = false },
                onResumeSelected = { gameViewModel.setButtonResume() },
                onEndSelected = { gameViewModel.stopGame() },
                onLeaveSelected = { navigateBack() }
            )
        }

        if (isGameEnded) {
            GameEndedDialog(
                onDismiss = { showPauseDialog = false },
                playerResult = playerResult,
                onLeaderboardSelected = { },
                onLeaveSelected = { navigateBack() }
            )
        }
    }
}